keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -99,182 +103,173 @@ File and block are first class citizens within the Kubernetes ecosystem.  Object
+1. `bucketMetaData`: stateful data relevant to the managing of the bucket but potentially inappropriate user knowledge (e.g. user's IAM role name)
+1. `phase`: is the current state of the BucketContent:
+    - `Bound`: the operator finished processing the request and bound the Bucket and BucketContent
+    - `Released`: the Bucket has been deleted, leaving the BucketContent unclaimed.
     - `Failed`: error and all retries have been exhausted.
     - `Retrying`: set when a recoverable driver or kubernetes error is encountered during bucket creation or access granting. Will be retried.
 
-
-\* Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+> \* Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
 
 #### BucketClass
 
-A cluster-wide scoped custom resource.
-During greenfield workflows an OBC references a Bucket Class (BC).  The bucket class defines a release policy, and specifies driver specific parameters, such as region, bucket lifecycle policies, etc., as well as the name of the driver as returned by the `GetDriverInfo()` rpc.  The driver name is used to filter OBs meant to be handled by the given driver.
+A cluster-scoped custom resource. The bucket class defines a release policy, and specifies driver specific parameters, such as region, bucket lifecycle policies, etc., as well as the name of the driver.  The driver name is used to filter BuckeContenst meant to be handled by a given driver.  In static bucket workflows, the driver name may be empty if the object bucket is defined.
 
 ```yaml
 apiVersion: cosi.io/v1alpha1
 kind: BucketClass
 metadata:
   name: 
-  namespace: [1]
-driver: [2]  # should this be named "provisioner:" ??
-config: string:string [3]
-releasePolicy: {"Delete", "Retain"} [4]
+provisioner: [1]
+config: string:string [2]
+releasePolicy: {"Delete", "Retain"} [3]
+targetBucket: [5]
+secretName: [6]
 ```
 
-1. `namespace`: BucketClasses are co-namespaced with their associated driver
-1. `driver`: Name of the driver, provided via the GetDriverInfo() rpc. Used to filter OBs.
+1. `provisioner`: Used by sidecars to filter BucketContent objects 
 1. `config`: object store specific key-value pairs passed to the driver.
-1.  `releasePolicy`: Prescribes outcome of an OBC/OB deletion.
+1.  `releasePolicy`: Prescribes outcome of a Deletion and Revoke events.
     - `Delete`:  the bucket and its contents are destroyed
     - `Retain`:  the bucket and its contents are preserved, only the user’s access privileges are terminated
-
-
+- `targetBucket`: A bucket connection endpoint of a brownfield or static bucket.  If provided, assumed to be a GrantAccess() operation.
```

Comment:
@jeffvance I took some liberty here to enable the workflows described for brownfield and static.  In writing them out, I realized that the workflows weren't congruent, even though the operations weren't that dissimilar. Regardless of brownfield or static, the admin has to define an endpoint.  But originally, that varied between the bucketClass and the bucketContent.

If I wanted to expose a static bucket to a cluster with 20 users who needed to connect to it, I'd have to define 20 BucketContent:Secret pairs AND track who I gave which BucketContent to.  Instead, I think it'd be much more user friendly to specify the user parent secret in the BC and let the COSI control automation handle Bucket:BucketContent binding and secret cloning.

Reply:
I think we want a many:1 relationship between bucket requests and a bucket class. Is there a way for static to also use a BC?

Reply:
I think its actually relatively simple if we enable the BC to specify the secret.  The COSI controller will generate bucketContents on demand as it does with brownfield.  This way we can expose a single static bucket, but dynamically distribute access to it.